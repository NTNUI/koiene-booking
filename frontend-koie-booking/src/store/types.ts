// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
