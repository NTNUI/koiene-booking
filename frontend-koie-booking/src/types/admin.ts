import { Component } from 'vue';

export interface AdminView {
  id: String; // Used to reference the buttons in SideBar
  title: String;
  icon: String;
  component: Component;
}
