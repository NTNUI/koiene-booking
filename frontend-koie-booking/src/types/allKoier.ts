export interface AllKoierData {
  koieTitle: string;
}
