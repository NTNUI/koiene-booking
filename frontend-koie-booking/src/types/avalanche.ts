import { TranslateResult } from 'vue-i18n';

export interface AvalancheData {
  result: string;
  levelFirstDay: string;
  levelSecondDay: string;
  levelThirdDay: string;
  firstDay: string;
  secondDay: string;
  thirdDay: string;
  fromDate: string;
  nextDate: string;
  lastDate: string;
  warningTypes: Array<TranslateResult>;
  colors: Array<string>;
}
