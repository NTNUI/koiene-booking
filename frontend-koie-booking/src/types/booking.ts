import { TranslateResult } from 'vue-i18n';

export interface BookingData {
  step: number;
  labels: Array<TranslateResult>;
}

export interface BookingFirstStepData {
  bedsTotal: number;
  bedsChosen: number;
  error: boolean;
}

export interface BookingSecondStepData {
  bedsChosen: number;
  bookingStep: number;
  guests: Array<Guest>;
  numberRules: object;
  nameRules: object;
  emailStrictRules: object;
  emailLooseRules: object;
  edited: boolean;
  validForm: boolean;
  memberPrice: number;
  nonMemberPrice: number;
}

export interface BookingThirdStepData {
  complete: boolean;
  stripeOptions: object;
  totalPrice: number;
}

export interface Guest {
  name: string;
  number: string;
  isMember: boolean;
  email?: string;
}
