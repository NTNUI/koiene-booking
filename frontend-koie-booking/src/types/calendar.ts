export interface CalendarData {
  menu1: boolean;
  menu2: boolean;
  beds: number;
  disabledDates: Array<String>;
  minimumDate: string;
  maximumDate: string;
  maximumDateTo: string;
  error: boolean;
  updatedDateFrom: boolean;
  temporaryDateTo: string;
  temporaryDateFrom: string;
  setDatesFinished: boolean;
}
