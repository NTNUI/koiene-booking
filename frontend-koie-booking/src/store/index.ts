import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
