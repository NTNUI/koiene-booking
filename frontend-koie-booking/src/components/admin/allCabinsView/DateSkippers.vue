<template>
  <v-row :justify="menuVersion === 'ALL' ? 'center' : menuVersion === 'NEGATIVE' ? 'start' : 'end'">
    <v-btn
      v-for="button in menuButtons"
      :key="button.label"
      :ref="button.label"
      class="navBtn"
      height="50px"
      width="50px"
      @click="button.method"
      >{{ button.label }}</v-btn
    >
  </v-row>
</template>

<script lang="ts">
import Vue from 'vue';
import dayjs from 'dayjs';
import store from '@/store/index';
import { addToDate } from '@/utils/dates';

interface SkipButton {
  label: string;
  method: () => void;
}

type MenuVersions = 'ALL' | 'NEGATIVE' | 'POSITIVE';

export default Vue.extend({
  name: 'DateSkippers',
  props: {
    menuVersion: {
      default: 'ALL',
      type: String,
    },
  },
  computed: {
    startDate() {
      return store.getters['adminBookings/getStartDate'];
    },
    menuButtons(): Array<SkipButton> {
      const buttons: Array<SkipButton> = [
        {
          label: '-1 å',
          method: () => this.skipDates(-1, 'year'),
        },
        {
          label: '-1 m',
          method: () => this.skipDates(-1, 'month'),
        },
        {
          label: '-1 u',
          method: () => this.skipDates(-1, 'week'),
        },
        {
          label: '-1 d',
          method: () => this.skipDates(-1, 'day'),
        },
        {
          label: '+1 d',
          method: () => this.skipDates(+1, 'day'),
        },
        {
          label: '+1 u',
          method: () => this.skipDates(+1, 'week'),
        },
        {
          label: '+1 m',
          method: () => this.skipDates(+1, 'month'),
        },
        {
          label: '+1 å',
          method: () => this.skipDates(+1, 'year'),
        },
      ];
      if (this.menuVersion === 'NEGATIVE') return buttons.slice(0, 4);
      else if (this.menuVersion === 'POSITIVE') return buttons.slice(4);
      else return buttons;
    },
  },
  methods: {
    skipDates(howMany: number, what: dayjs.OpUnitType) {
      const startDate = addToDate(this.startDate, howMany, what);
      store.commit('adminBookings/setStartDate', startDate);
      const endDate = addToDate(startDate, 7, 'day');
      store.dispatch('adminBookings/MOUNT_CABINS_WITH_BOOKINGS', { startDate: startDate, endDate: endDate });
    },
  },
});
</script>

<style scoped></style>
