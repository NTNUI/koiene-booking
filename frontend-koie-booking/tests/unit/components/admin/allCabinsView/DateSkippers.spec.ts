import Vue, { VueConstructor } from 'vue';
import Vuetify, { Vuetify as VuetifyType } from 'vuetify';
import scssVars from '@/styles/variables.scss';
import mockAxios from 'jest-mock-axios';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue, shallowMount } from '@vue/test-utils';

// Components or views
import DateSkippers from '@/components/admin/allCabinsView/DateSkippers.vue';
import dayjs from 'dayjs';
import { Store } from 'vuex';
import { RootState } from '@/store/types';

describe('Component DateSkippers.vue', () => {
  let localVue: VueConstructor<Vue>;
  let vuetify: VuetifyType;
  let store: Store<RootState>;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(Vuetify);
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Mounts the component', () => {
    const wrapper = mount(DateSkippers);
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  it('Matches snapshot', () => {
    const wrapper = mount(DateSkippers);
    expect(wrapper).toMatchSnapshot();
  });

  it('Renders only negative menu', () => {
    const wrapper = mount(DateSkippers, { propsData: { menuVersion: 'NEGATIVE' } });
    const labels = ['-1 책', '-1 m', '-1 u', '-1 d'];
    for (const label of labels) {
      const button = wrapper.find({ ref: label });
      expect(button.exists()).toBeTruthy();
    }
  });

  it('Renders only positive menu', () => {
    const wrapper = mount(DateSkippers, { propsData: { menuVersion: 'NEGATIVE' } });
    const labels = ['+1 책', '+1 m', '+1 u', '+1 d'];
    for (const label of labels) {
      const button = wrapper.find({ ref: label });
      expect(button.exists()).toBeTruthy();
    }
  });

  it('Skip button calls skipDates with correct argument', () => {
    const skipDates = jest.fn();
    const wrapper = mount(DateSkippers, {
      localVue,
      vuetify,
      methods: {
        skipDates: skipDates,
      },
      propsData: { menuVersion: 'ALL' },
    });
    const expectedArguments: { label: string; howMany: number; what: dayjs.OpUnitType }[] = [
      {
        label: '-1 책',
        howMany: -1,
        what: 'year',
      },
      {
        label: '-1 m',
        howMany: -1,
        what: 'month',
      },
      {
        label: '-1 u',
        howMany: -1,
        what: 'week',
      },
      {
        label: '-1 d',
        howMany: -1,
        what: 'day',
      },
      {
        label: '+1 d',
        howMany: +1,
        what: 'day',
      },
      {
        label: '+1 u',
        howMany: +1,
        what: 'week',
      },
      {
        label: '+1 m',
        howMany: +1,
        what: 'month',
      },
      {
        label: '+1 책',
        howMany: +1,
        what: 'year',
      },
    ];

    for (const argument of expectedArguments) {
      const button = wrapper.find({ ref: argument.label });
      button.trigger('click');
      expect(skipDates).toHaveBeenLastCalledWith(argument.howMany, argument.what);
    }
  });
});
