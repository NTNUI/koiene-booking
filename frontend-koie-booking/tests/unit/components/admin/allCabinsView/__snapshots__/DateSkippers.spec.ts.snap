// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component DateSkippers.vue Matches snapshot 1`] = `<div class="row justify-center"><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">-1 å</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">-1 m</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">-1 u</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">-1 d</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">+1 d</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">+1 u</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">+1 m</span></button><button type="button" class="navBtn v-btn v-btn--contained theme--light v-size--default" style="height: 50px; width: 50px;"><span class="v-btn__content">+1 å</span></button></div>`;
